kcm: add support for kerberos tgt renewals

justin-stephenson opened the pull request:
This allows configuration of KCM to attempt renewals for renew-applicable kerberos tickets stored in kcm, such as a TGT retrieved with `kinit` on the command-line.

krb5* renewal and lifetime options are configurable in the [kcm] section, if they do not exist in the [kcm] section then we check and fallback to using the first `auth_provider=krb5` domain in sssd.conf.

This support is only added to the secdb ccache backend. The overall high-level logic used here is similar to the existing pam_sss krb5 renewal code, adding necessary changes to unmarshal and retrieve ticket information from KCM secrets db.

Renewal is only attempted after half of the tgt lifetime has been reached.

frozencemetery commented:
(I'll want to review this once it passes CI.)

justin-stephenson commented:
Latest CI failures are unrelated.

frozencemetery reviewed (changes requested):
I've started discussion upstream in krb5 about making public functions for
cred marshalling in krb5.  This would allow SSSD to avoid reimplementing
several hundred lines of parser logic to produce a krb5_creds object - which
seems preferable to me.

In the interest of providing other feedback, though, I've reviewed the code as
it stands - noting as usual that I'm not reviewing talloc logic and deep
SSSD-isms.  There are several persistent problems that come up frequently
enough that I think it's clearer to call out here rather than marking each
instance.

Style violations (So that we're all on the same page, I'm referring to SSSD
coding style: https://sssd.io/docs/developers/coding_style.html )

- Weird whitespace - it's easy to spot these reading through the github viewer
  because things don't line up.  Coding style says "MUST: No tabs all
  indentation 4 spaces."  (Presumably this isn't true for Makefiles, which are
  required to be tabs; there are some problems there too.)

- line lengths - While it's not a "MUST", < 80 character lines are "HIGHLY
  RECOMMENDED".

- goto labels: I believe the use of "immediate" should be replaced with
  "done"/"fail".

- Multi-line comments "MUST" have `/*` and `*/` on their own lines, as well as
  being "real sentences" and resembling "real paragraphs".

- Mid-function variable declaration - "MUST: Always declare variables at the
  top of the function or block."

General clarity things:

- `ret = foo; return ret;`.  There's no need for this; just `return ret`.

- `return foo; done: return foo;`.  The double return is redundant, and
  moreover the label isn't necessary at all; just return directly within the
  function.

Further comments are inline.

frozencemetery commented:
(krb5 discussion PR: https://github.com/krb5/krb5/pull/1153 )

justin-stephenson commented on src/tests/intg/krb5utils.py:
```diff
@@ -62,6 +62,11 @@ def kinit(self, principal, password, env=None):
         args = ["kinit", principal]
         return self._run_in_env(args, password.encode('utf-8'), env)
 
+    def kinit_with_opts(self, principal, password, options, env=None):
```
Fixed, thanks.

justin-stephenson commented on src/tests/intg/test_kcm.py:
```diff
@@ -585,3 +627,40 @@ def test_kcm_secrets_quota(setup_for_kcm_sec,
+    initial_times = testenv.k5util.list_times()
+
+    # Wait for renewal to trigger once, after renew interval
+    time.sleep(15)
```
It's not ideal, but the same approach is used for other integration tests.

pbrezina replied:
`libfaketime` looks interesting. It would be good if it helps to reduce the test run time. But it is probably out of scope of this work, unless you'd like to look at it.

justin-stephenson commented on src/tests/cmocka/test_kcm_renewals.c:
```diff
@@ -0,0 +1,299 @@
+    if (mask) {
+        st_mode = stat_buf->st_mode & mask;
+    } else {
+        st_mode = stat_buf->st_mode & (S_IFMT|ALLPERMS);
```
This is the same code from `src/util/check_and_open.c` so I would prefer to keep it the same

frozencemetery replied:
If it's the same code, shouldn't it not be copied around? :)

justin-stephenson replied:
Yes definitely, but this was the easiest way to avoid including several different modules and libs into the test just to wrap the fstat call

justin-stephenson commented on src/tests/cmocka/test_kcm_renewals.c:
```diff
@@ -0,0 +1,299 @@
+        DEBUG(SSSDBG_TRACE_LIBS,
+              "File has the wrong (bit masked) mode [%.7o], "
+              "expected [%.7o].\n",
+              (st_mode & ALLPERMS), (mode & ALLPERMS));
```
Same as above comment

justin-stephenson commented on src/tests/cmocka/test_utils.c:
```diff
@@ -2041,6 +2048,138 @@ static void test_sss_filter_sanitize_dn(void **state)
+    krb5_principal client_principal;
+    krb5_principal server_principal;
+    static krb5_address addr;
+    int add=0x12345;
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -60,16 +60,184 @@ struct auth_data {
+    *_req = krreq;
+    return EOK;
+fail:
+    if (ret != EOK) {
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -60,16 +60,184 @@ struct auth_data {
+        goto fail;
+    }
+
+    /* FIXME: Handle old ccname correctly? */
```
Fixed.

justin-stephenson commented on src/providers/krb5/krb5_child_handler.c:
```diff
@@ -138,15 +138,20 @@ static errno_t create_send_buffer(struct krb5child_req *kr,
-        break;
-    default:
-        return EINVAL;
+    /* Renewals from KCM do not initialize kr->dom  */
+    if (kr->pd->cmd == SSS_CMD_RENEW) {
+            posix_domain = 1;
+    } else {
+        switch (kr->dom->type) {
+        case DOM_TYPE_POSIX:
+            posix_domain = 1;
+            break;
+        case DOM_TYPE_APPLICATION:
+            posix_domain = 0;
+            break;
+        default:
+            return EINVAL;
+        }
```
Updated, thanks.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -58,6 +58,86 @@ struct auth_data {
+static void kcm_renew_tgt(struct tevent_context *ev, struct tevent_timer *te,
+                      struct timeval current_time, void *private_data)
+{
+    struct auth_data *auth_data = talloc_get_type(private_data,
+                                                  struct auth_data);
+    struct tevent_req *req;
```
These were placeholders used during testing, the commit  `KCM: Execute krb5 child to perform renewal ` adds to `kcm_renew_tgt` and `kcm_renew_tgt_done`. I can cleanup the git history here if you would prefer.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -58,6 +58,86 @@ struct auth_data {
+        DEBUG(SSSDBG_TRACE_INTERNAL,
+              "Checking [%s] for renewal at [%.24s].\n", renew_data->ccname,
+                  ctime(&renew_data->start_renew_at));
+        if (renew_data->start_renew_at <= now) {
```
Sorry, I don't quite understand what you mean, and what is the advantage?

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -80,6 +160,8 @@ static void kcm_renew_tgt_timer_handler(struct tevent_context *ev,
         return;
     }
 
+    ret = kcm_renew_all_tgts(renew_tgt_ctx);
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,215 @@
+    struct renew_data *renew_data = NULL;
+
+    if (krb5_ctx->renew_tgt_ctx == NULL) {
+        DEBUG(SSSDBG_TRACE_LIBS ,"Renew context not initialized, "
```
Where is the typo in  `Renew context not initialized, automatic renewal not available.`

frozencemetery replied:
After `SSSDBG_TRACE_LIBS`, comma goes before space.

justin-stephenson replied:
Thanks, fixed.

justin-stephenson commented on src/responder/kcm/kcmsrv_ccache_secdb.c:
```diff
@@ -577,6 +580,193 @@ static errno_t ccdb_secdb_init(struct kcm_ccdb *db,
+
+	cli_cred = talloc_zero(ev, struct cli_creds);
+    if (cli_cred == NULL) {
+        ret = ENOMEM;
```
Yes, fixed.

justin-stephenson commented on src/responder/kcm/kcm.c:
```diff
@@ -83,10 +86,242 @@ static errno_t kcm_get_ccdb_be(struct kcm_ctx *kctx)
+                                CONFDB_DOMAIN_AUTH_PROVIDER,
+                                0, &auth_provider);
+
+        if (auth_provider != NULL) {
```
I don't really prefer this style, but if another developer agrees I'm happy to fix it.

simo5 replied:
agree with @frozencemetery

justin-stephenson replied:
Fixed.

justin-stephenson commented on src/responder/kcm/kcm.c:
```diff
@@ -96,6 +331,7 @@ static int kcm_get_config(struct kcm_ctx *kctx)
     if (ret != EOK) {
         DEBUG(SSSDBG_FATAL_FAILURE,
               "Failed to get file descriptors limit\n");
+        talloc_free(krb5_ctx);
```
Yes, fixed.

frozencemetery commented:
Hi, you've written "fixed" below several things, but the code isn't change.  Are you missing a push?

justin-stephenson commented:
> I've started discussion upstream in krb5 about making public functions for
> cred marshalling in krb5. This would allow SSSD to avoid reimplementing
> several hundred lines of parser logic to produce a krb5_creds object - which
> seems preferable to me.
> 
> In the interest of providing other feedback, though, I've reviewed the code as
> it stands - noting as usual that I'm not reviewing talloc logic and deep
> SSSD-isms. There are several persistent problems that come up frequently
> enough that I think it's clearer to call out here rather than marking each
> instance.
> 
> Style violations (So that we're all on the same page, I'm referring to SSSD
> coding style: https://sssd.io/docs/developers/coding_style.html )
> 
>     * Weird whitespace - it's easy to spot these reading through the github viewer
>       because things don't line up.  Coding style says "MUST: No tabs all
>       indentation 4 spaces."  (Presumably this isn't true for Makefiles, which are
>       required to be tabs; there are some problems there too.)
> 
>     * line lengths - While it's not a "MUST", < 80 character lines are "HIGHLY
>       RECOMMENDED".
> 
>     * goto labels: I believe the use of "immediate" should be replaced with
>       "done"/"fail".
> 
>     * Multi-line comments "MUST" have `/*` and `*/` on their own lines, as well as
>       being "real sentences" and resembling "real paragraphs".
> 
>     * Mid-function variable declaration - "MUST: Always declare variables at the
>       top of the function or block."
> 
> 
> General clarity things:
> 
>     * `ret = foo; return ret;`.  There's no need for this; just `return ret`.
> 
>     * `return foo; done: return foo;`.  The double return is redundant, and
>       moreover the label isn't necessary at all; just return directly within the
>       function.
> 
> 
> Further comments are inline.

Thanks very much for the review. I have addressed most of the in-line comments, but I still need to fix the style issues you mention here in this specific comment.

I'll push to the PR again once done.

justin-stephenson commented:
I addressed several of the aforementioned issues, I'll make a second pass tomorrow. I could not find any instances of `Mid-function variable declaration` but maybe I missed it when looking over the changes.

Thanks again for the review and also for submitting a request to make the marshalling/unmarshalling functions public!

frozencemetery commented:
The krb5 PR has merged upstream.  For convenience, I've backported the two functions to Fedora rawhide starting in krb5-1.19-0.beta2.3.fc34.

sumit-bose commented:
> The krb5 PR has merged upstream. For convenience, I've backported the two functions to Fedora rawhide starting in krb5-1.19-0.beta2.3.fc34.

Hi Robbie,

thanks for the effort. What would be your suggestion for the way forward? We can add a configure check if the new functions are already available. But if they are not available would you recommend to fall-back to the SSSD implementation of the functions or to build sssd_kcm without the certificate renewal feature?

bye,
Sumit

frozencemetery commented:
> thanks for the effort. What would be your suggestion for the way forward? We can add a configure check if the new functions are already available. But if they are not available would you recommend to fall-back to the SSSD implementation of the functions or to build sssd_kcm without the certificate renewal feature?

(Assuming this is a typo for "credential renewal".)

My preference would be to not build the functionality.  If there's an older release you need this functionality on, I can probably backport it.

The point is to allow sssd-kcm to not carry code for {de,}serializing the credentials objects.  It reduces testing burden, and it makes debugging for us easier in the event that there are any issues.

sumit-bose commented:
> > thanks for the effort. What would be your suggestion for the way forward? We can add a configure check if the new functions are already available. But if they are not available would you recommend to fall-back to the SSSD implementation of the functions or to build sssd_kcm without the certificate renewal feature?
> 
> (Assuming this is a typo for "credential renewal".)

yes :-)

> 
> My preference would be to not build the functionality. If there's an older release you need this functionality on, I can probably backport it.
> 
> The point is to allow sssd-kcm to not carry code for {de,}serializing the credentials objects. It reduces testing burden, and it makes debugging for us easier in the event that there are any issues.

Ok, fair point. @justin-stephenson, if you agree with this as well please let me know if you need help with the configure/automake related changes?

bye,
Sumit

justin-stephenson commented:
Thank you, I'll work on it and update the PR.

justin-stephenson commented:
Sorry for the delay. I updated the PR to

 * Use the exported krb5 marshalling credentials functions
 * Build KCM renewals code conditionally, if the krb5 marshalling functions are available then we auto-detect that and build with KCM renewals, explicit `--enable-kcm-renewal` and `--disable-kcm-renewal` ./configure options can also be provided to override auto-detection.

I also noticed that the responder idle timeout can shutdown KCM when renewals are configured, so I added commit b206ba3c0340877b0e6df2e530fdb350b838ac5d to disable the responder idle timeout when renewals code is built. Once KCM comes up with renewals configured it will stay active.

This does not handle the case when a renewal is expected to occur after the system boots but before any kerberos activity has occurred to socket-activate KCM. I don't know if this is a valid case we need to handle, it seems unlikely to me but @sumit-bose suggested we could install a .timer file for sssd-kcm.

justin-stephenson commented:
CI fails as expected with `configure: error: krb5 marshalling functions not available, --disable-kcm-renewal should be used`

frozencemetery reviewed (changes requested):
I don't see a callsite for `krb5_marshal_credentials()` - if you don't call it, you shouldn't be checking for it at build time.

I don't expect the project to merge this without passing CI.  To that end, it might be worth considering whether kcm-renewal should have fallback at build-time if it doesn't find the function.  Additionally, the code is present in the (branched) fc34 as well as rawhide (fc35) - sssd could consider upgrading what fedora version is used.

I do not have capacity to provide a detailed code review except where it touches krb5 and highly suggest someone else do so before merging :)

sumit-bose commented:
Hi,

@justin-stephenson, you can add `--disable-kcm-renewal` for platforms where the calls are not available in `contrib/ci/configure.sh`

@pbrezina, can you update the rawhide vagrant image and add fedora34? Thanks.

bye,
Sumit

justin-stephenson commented:
Changes made as requested.

justin-stephenson commented:
@sumit-bose How do I conditionally exclude the `test_kcm_renewals` integration test from `src/tests/intg/test_kcm.py` when kcm renewals are not being built? See https://s3.eu-central-1.amazonaws.com/sssd-ci/PR-5450/10/fedora32/ci-make-intgcheck.log

justin-stephenson commented:
Disregard my previous comment, I added a custom pytest marker to skip the intg test if not built with renewal support

justin-stephenson commented:
It looks like most the CI failures are unrelated to this PR, @sumit-bose could you please confirm this?

sumit-bose commented:
Hi,

thanks for updating the patches so that the build works automatically on platforms without the new libkrb5 calls. The CI failures with rawhide and F34 are (unfortunately) currently expected.

bye,
Sumit

pbrezina commented on src/man/sssd-kcm.8.xml:
```diff
@@ -249,6 +280,19 @@ systemctl restart sssd-kcm.service
+                        Default: 0 (Automatic renewal is disabled)
+                    </para>
+                </listitem>
+            </varlistentry>
```
Remaining options are undocumented intentionally?

justin-stephenson replied:
I tried to keep the man page more concise. Above I provide the configurable krb5_* options list and state that sssd-krb5 contains the descriptions of each.

If you prefer I am happy to document each option in this man page.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+    struct sss_domain_info *domains = NULL;
+    struct sss_domain_info *dom = NULL;
+
+   ret = confdb_get_domains(kctx->rctx->cdb, &domains);
```
Wrong indentation?

justin-stephenson replied:
Fixed.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+        }
+    }
+
+    return EOK;
```
ret = EOK

justin-stephenson replied:
Fixed.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+                                0, &auth_provider);
+
+        if (auth_provider != NULL) {
+            if (strcasecmp(auth_provider, "krb5") == 0) {
```
This needs to be at least documented.

I'm unsure about this functionality, it will cause troubles in multi-domain setup. Ideally we should match the ccache to a domain, then read configuration from the domain. But since this would be too difficult, imho it is best to allow configuration only in [kcm] instead of trying to be smart. (I did not go through all the comments here, I apologize if it was already discussed somewhere).

justin-stephenson replied:
Okay thanks for the input, it was a recommendation by @sumit-bose - Sumit do you have strong opposition to removing the fallback to `auth_provider = krb5` options for KCM renewals?

justin-stephenson replied:
I updated the man page to ensure that the fallback any existing auth_provider=krb5 is documented.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+            /* Not fatal */
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to read auth provider options\n");
+        }
+     }
```
Bad indentation.

justin-stephenson replied:
Fixed.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+        }
+     }
+
+    return EOK;
```
ret = EOK

justin-stephenson replied:
Fixed.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -294,4 +385,165 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+	/* Add any renew-applicable KCM tickets to renew table */
+	ret = kcm_ccdb_renew_init(renew_tgt_ctx->rctx, renew_tgt_ctx->krb5_ctx,
+                              ev, renew_tgt_ctx->db);
+    if (ret != 0) {
```
ret != EOK

justin-stephenson replied:
Fixed.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -294,4 +385,165 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+    }
+
+    next = tevent_timeval_current_ofs(renew_tgt_ctx->timer_interval, 0);
+    renew_tgt_ctx->te = tevent_add_timer(ev, renew_tgt_ctx,
```
Check if the timer was created (te != NULL)

justin-stephenson replied:
Done.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -294,4 +385,165 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+     * discard_const_p here.
+     */
+    key.type = HASH_KEY_STRING;
+    key.str = discard_const_p(char, upn);
```
Did you consider using `sss_ptr_hash` interface instead?

justin-stephenson replied:
No I did not, is that interface preferred?

pbrezina replied:
It does little bit more work to make sure the hash table stays consistent (does not contain values that were freed, frees values when you delete them, handle duplicate values, etc.) which may or may not be needed in this case. It is also little bit easier to use when you manipulate entries. I would say it is preferred.

At the very least, I don't see you `free(value)` when you `hash_delete(key)`.

justin-stephenson replied:
Thanks for the suggestion, I updated the code to use the `sss_ptr_hash` interface.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -385,6 +387,84 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+        renew_data = talloc_get_type(entries[c].value.ptr, struct renew_data);
+        DEBUG(SSSDBG_TRACE_INTERNAL,
+              "Checking [%s] for renewal at [%.24s].\n", renew_data->ccname,
+                  ctime(&renew_data->start_renew_at));
```
Move the debug string to the first line, then you can fit it to two lines.

justin-stephenson replied:
Fixed.

pbrezina commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -385,6 +387,84 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+                                          kcm_renew_tgt, auth_data);
+                    if (te == NULL) {
+                        DEBUG(SSSDBG_CRIT_FAILURE,
+                              "tevent_add_timer failed.\n");
```
Can this fit to one line?

justin-stephenson replied:
Yes fixed.

pbrezina reviewed (changes requested):
Hi Justin, I have just few minor comments. You did a good job on this task.

pbrezina commented on src/responder/kcm/kcmsrv_ccache_secdb.c:
```diff
@@ -577,6 +580,206 @@ static errno_t ccdb_secdb_init(struct kcm_ccdb *db,
+
+immediate:
+    if (krb_context != NULL) {
+        krb5_free_context(krb_context);
```
krb_context may not be initialized here

pbrezina reviewed (changes requested):
There are few minor comments... but:

If I understand it correctly, this functionality requires KCM to be running. KCM is currently socket activated so I see two problems:

1. Unless somebody use Kerberos regularly, keeping the KCM busy, the renewal will not work. So this makes the feature unfortunately useless, since you want to renew the ticket mostly when you don't use the computer for longer period of times, e.g. when you have session locked during a weekend.
    - the timer will be always scheduled in a destined future time when KCM is started, but KCM will likely terminate before we get even close to this time (idle timeout is five minutes). 
2. You add creds to renew table on two places:
    a) when KCM process starts `kcm_process_init`
    b) when renew timer is triggers `kcm_renew_tgt_timer_handler`
  However, since it is socket activated, b) is very unlikely to happen. And we already have performance issues so its probably not a very good idea to do it in a).

The code itself is fine, but unless I am missing something, it is currently unusable. We have to either avoid socket activation, which is not desirable. Or find a way how to execute the process periodically in certain intervals (systemd timer might help here) and change the renew table logic.

pbrezina commented:
> There are few minor comments... but:
> 
> If I understand it correctly, this functionality requires KCM to be running. KCM is currently socket activated so I see two problems:
> 
>     1. Unless somebody use Kerberos regularly, keeping the KCM busy, the renewal will not work. So this makes the feature unfortunately useless, since you want to renew the ticket mostly when you don't use the computer for longer period of times, e.g. when you have session locked during a weekend.
>        
>        * the timer will be always scheduled in a destined future time when KCM is started, but KCM will likely terminate before we get even close to this time (idle timeout is five minutes).
> 
>     2. You add creds to renew table on two places:
>        a) when KCM process starts `kcm_process_init`
>        b) when renew timer is triggers `kcm_renew_tgt_timer_handler`
>        However, since it is socket activated, b) is very unlikely to happen. And we already have performance issues so its probably not a very good idea to do it in a).
> 
> 
> The code itself is fine, but unless I am missing something, it is currently unusable. We have to either avoid socket activation, which is not desirable. Or find a way how to execute the process periodically in certain intervals (systemd timer might help here) and change the renew table logic.

Ah, I missed the last patch: `KCM: Disable responder idle timeout with renewals`. So it will work correclty. But I wonder if it would be better to keep the idle timeout enabled. What we could do is to make systemd timer send a SSSD-specific KCM op code periodically and renew the tickets per-request. This would also simplify the logic by a lot since you would not have to keep the hash table and timers.

alexey-tikhonov commented:
> But I wonder if it would be better to keep the idle timeout enabled. 

What's wrong with keeping an idle process "running"? Sleeping process with small memory footprint shouldn't have any resource implications, right?

pbrezina commented:
> > But I wonder if it would be better to keep the idle timeout enabled.
> 
> What's wrong with keeping an idle process "running"? Sleeping process with small memory footprint shouldn't have any resource implications, right?

It's not sleeping, it still spins in tevent loop doing stuff which may have a negative impact on battery. The impact from a single process may not be significant, but there are lots of such processes so if it can be avoided it is certainly welcomed. However, there's nothing wrong about it per say. It just makes sense to me to keep KCM a short lived service, especially since it will also reduce the amount of code that is required to provide the functionality.

alexey-tikhonov commented:
> It's not sleeping, it still spins in tevent loop doing stuff which may have a negative impact on battery. 

That's exactly my question: what is it doing? IIUC, it should be sleeping on `epoll()` (say 99.999% of the time)

If it actually does something useful, it means process would have to be socket activated otherwise which is much more expensive than awaking from epoll().
If it doesn't do anything usefull, then what is it doing?

justin-stephenson commented:
> Ah, I missed the last patch: `KCM: Disable responder idle timeout with renewals`. So it will work correclty. But I wonder if it would be better to keep the idle timeout enabled. What we could do is to make systemd timer send a SSSD-specific KCM op code periodically and renew the tickets per-request. This would also simplify the logic by a lot since you would not have to keep the hash table and timers.

I'm fine with this approach, but if the systemd timer file is installed conditionally at build time(if KCM renewals are built), then what interval value, i.e. amount of time that KCM wakes up to attempt renewals, should we set in the systemd timer file? Currently the renew interval is defined with the `krb5_renew_interval` option in sssd.conf. I suppose the other side effect is that falllback to `auth_provider=krb5` renew config options would no longer work.

pbrezina commented:
> > It's not sleeping, it still spins in tevent loop doing stuff which may have a negative impact on battery.
> 
> That's exactly my question: what is it doing? IIUC, it should be sleeping on `epoll()` (say 99.999% of the time)
> 
> If it actually does something useful, it means process would have to be socket activated otherwise which is much more expensive than awaking from epoll().
> If it doesn't do anything usefull, then what is it doing?

epoll is just one of the tevent mechanisms, there is much more to it. It does not check only epoll, but also go into internal structures to watch for signals and to trigger timed events and tevent reqs. But lets not dive into it, I formulated my previous answer wrongly and made battery life a stronger point then I meant. If this was the reason to keep it as short lived process it is certainly something then needs to be measured.

My point was that lots of logic that Justin introduced would not be necessary.







> > Ah, I missed the last patch: `KCM: Disable responder idle timeout with renewals`. So it will work correclty. But I wonder if it would be better to keep the idle timeout enabled. What we could do is to make systemd timer send a SSSD-specific KCM op code periodically and renew the tickets per-request. This would also simplify the logic by a lot since you would not have to keep the hash table and timers.
> 
> I'm fine with this approach, but if the systemd timer file is installed conditionally at build time(if KCM renewals are built), then what interval value, i.e. amount of time that KCM wakes up to attempt renewals, should we set in the systemd timer file? Currently the renew interval is defined with the `krb5_renew_interval` option in sssd.conf. This is an important consideration because if the renewal interval is too high then we could miss renewing tickets that have already expired, too low and it may add unnecessary KCM load.
> 
> I suppose the other side effect is that falllback to `auth_provider=krb5` renew config options would no longer work.

Fair point. Let's continue with this patch set as is. Since SSSD is not currently relying that much on systemd it is probably a better choice. We can revisit this if we ever switch from monitor to systemd.

pbrezina commented on src/responder/kcm/kcm.c:
```diff
@@ -140,6 +140,7 @@ static int kcm_get_config(struct kcm_ctx *kctx)
         goto done;
     }
 
+#ifndef HAVE_KCM_RENEWAL
```
This will disable idle timeout during build time. Wouldn't it be better to disable it only if renewals are enabled in configuration?

justin-stephenson replied:
You are correct, new version implements this behaviour.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+static int kcm_get_auth_provider_options(struct kcm_ctx *kctx,
+                                         struct krb5_ctx *krb5_ctx)
+{
+    errno_t ret = 0;
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+    const char *conf_path = NULL;
+    char *auth_provider = NULL;
+    struct sss_domain_info *domains = NULL;
+    struct sss_domain_info *dom = NULL;
```
Good to know, thanks. Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+                /* Validate */
+                validate = dp_opt_get_bool(opts, KRB5_VALIDATE);
+                ret = dp_opt_set_bool(krb5_ctx->opts, KRB5_VALIDATE, validate);
+                if (ret != 0) {
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -294,4 +386,161 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+    renew_data->renew_till = tgtt->renew_till;
+
+    /* Overwrite existing entry with updated times */
+    if (sss_ptr_hash_has_key(krb5_ctx->renew_tgt_ctx->tgt_table, upn)) {
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -386,6 +388,81 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+        if (renew_data->renew_till < now) {
+            DEBUG(SSSDBG_TRACE_INTERNAL, "Renew time exceeded, removing [%s].\n",
+                                         renew_data->ccname);
+            sss_ptr_hash_delete(renew_tgt_ctx->tgt_table, keys[i].c_str, true);
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -388,16 +402,170 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+        goto done;
+    }
+    ret = parse_krb5_child_response(ctx, ctx->buf, ctx->len, ctx->kr->pd,
+            0, &res);
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+                                kctx->rctx,
+                                conf_path,
+                                CONFDB_DOMAIN_AUTH_PROVIDER,
+                                0, &auth_provider);
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+                            kctx->rctx,
+                            kctx->rctx->confdb_service_path,
+                            CONFDB_KCM_KRB5_LIFETIME,
+                            0, &lifetime_str);
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -0,0 +1,297 @@
+                            kctx->rctx,
+                            kctx->rctx->confdb_service_path,
+                            CONFDB_KCM_KRB5_RENEWABLE_LIFETIME,
+                            0, &rtime);
```
Fixed.

justin-stephenson commented:
I pushed a new version addressing the latest review items from Pavel. The CI failure `Details` link shows no actual files, just an empty header. Is it expected?

justin-stephenson commented on src/responder/kcm/kcmsrv_ccache_secdb.c:
```diff
@@ -577,6 +580,193 @@ static errno_t ccdb_secdb_init(struct kcm_ccdb *db,
+        goto immediate;
+    }
+    /* Extract credentials */
+    if (cc != NULL) {
```
Fixed.

pbrezina reviewed (changes requested):
Hi Justin,
I'm sorry it takes so long, but there are few more thinks. They are mostly minor, but one makes kcm fail to start.

Warning:

```
/home/pbrezina/workspace/sssd/src/responder/kcm/kcmsrv_ccache_secdb.c: In function ‘renew_check_ccaches’:
/home/pbrezina/workspace/sssd/src/responder/kcm/kcmsrv_ccache_secdb.c:728:15: error: ‘state’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
  728 |         ret = secdb_container_url_req(state, sctx, cli_cred, &sreq);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

If the cache contains uid that is not resolvable then kcm fails to start:

```
[root /var/log/sssd]# /usr/libexec/sssd/sssd_kcm --uid 0 --gid 0 --debug-level 0xfff0
(2021-03-25 12:12:49:260824): [sssd] [become_user] (0x0200): Trying to become user [0][0].
(2021-03-25 12:12:49:260883): [sssd] [become_user] (0x0200): Already user [0].
(2021-03-25 12:12:49:263412): [kcm] [ldb] (0x0400): server_sort:Unable to register control with rootdse!
(2021-03-25 12:12:49): [kcm] [server_setup] (0x0040): Starting with debug level = 0xfff0
(2021-03-25 12:12:49): [kcm] [server_setup] (0x0400): CONFDB: /var/lib/sss/db/config.ldb
(2021-03-25 12:12:49): [kcm] [kcm_get_ccdb_be] (0x0100): KCM database type: secdb
(2021-03-25 12:12:49): [kcm] [kcm_ccdb_init] (0x0200): KCM back end: libsss_secrets
(2021-03-25 12:12:49): [kcm] [ccdb_secdb_init] (0x2000): secdb initialized
(2021-03-25 12:12:49): [kcm] [sss_sec_list_cc_uids] (0x2000): uid: [91600000]
(2021-03-25 12:12:49): [kcm] [sss_sec_list_cc_uids] (0x2000): uid: [1000]
(2021-03-25 12:12:49): [kcm] [ccdb_secdb_renew_init] (0x2000): Found [2] ccache uids
(2021-03-25 12:12:49): [kcm] [renew_check_ccaches] (0x0040): Failed to get pwd entry for [91600000]
(2021-03-25 12:12:49): [kcm] [ccdb_secdb_renew_init] (0x0040): Error checking ccaches in secdb
(2021-03-25 12:12:49): [kcm] [kcm_ccdb_renew_init] (0x0020): Failure to execute ccdb renewal init
(2021-03-25 12:12:49): [kcm] [kcm_process_init] (0x0010): fatal error initializing KCM ccdb renewals
(2021-03-25 12:12:49): [kcm] [kcm_responder_ctx_destructor] (0x0400): Responder is being shut down
```

justin-stephenson commented:
Thanks a lot Pavel for the further review, it is better to get everything resolved now than having to fix issues later.

> If the cache contains uid that is not resolvable then kcm fails to start:
> 
> ```
> [root /var/log/sssd]# /usr/libexec/sssd/sssd_kcm --uid 0 --gid 0 --debug-level 0xfff0
> (2021-03-25 12:12:49:260824): [sssd] [become_user] (0x0200): Trying to become user [0][0].
> (2021-03-25 12:12:49:260883): [sssd] [become_user] (0x0200): Already user [0].
> (2021-03-25 12:12:49:263412): [kcm] [ldb] (0x0400): server_sort:Unable to register control with rootdse!
> (2021-03-25 12:12:49): [kcm] [server_setup] (0x0040): Starting with debug level = 0xfff0
> (2021-03-25 12:12:49): [kcm] [server_setup] (0x0400): CONFDB: /var/lib/sss/db/config.ldb
> (2021-03-25 12:12:49): [kcm] [kcm_get_ccdb_be] (0x0100): KCM database type: secdb
> (2021-03-25 12:12:49): [kcm] [kcm_ccdb_init] (0x0200): KCM back end: libsss_secrets
> (2021-03-25 12:12:49): [kcm] [ccdb_secdb_init] (0x2000): secdb initialized
> (2021-03-25 12:12:49): [kcm] [sss_sec_list_cc_uids] (0x2000): uid: [91600000]
> (2021-03-25 12:12:49): [kcm] [sss_sec_list_cc_uids] (0x2000): uid: [1000]
> (2021-03-25 12:12:49): [kcm] [ccdb_secdb_renew_init] (0x2000): Found [2] ccache uids
> (2021-03-25 12:12:49): [kcm] [renew_check_ccaches] (0x0040): Failed to get pwd entry for [91600000]
> (2021-03-25 12:12:49): [kcm] [ccdb_secdb_renew_init] (0x0040): Error checking ccaches in secdb
> (2021-03-25 12:12:49): [kcm] [kcm_ccdb_renew_init] (0x0020): Failure to execute ccdb renewal init
> (2021-03-25 12:12:49): [kcm] [kcm_process_init] (0x0010): fatal error initializing KCM ccdb renewals
> (2021-03-25 12:12:49): [kcm] [kcm_responder_ctx_destructor] (0x0400): Responder is being shut down
> ```

Is it valid to treat renewal failures as not fatal, and return EOK from `kcm_ccdb_renew_init` instead of current behavior shown below? If not what is the preferred way to handle this?

```
647     /* Add any renew-applicable KCM tickets to renew table */
648     ret = kcm_ccdb_renew_init(renew_tgt_ctx->rctx, renew_tgt_ctx->krb5_ctx,
649                               ev, renew_tgt_ctx->db);
650     if (ret != EOK) {
651         DEBUG(SSSDBG_CRIT_FAILURE, "Failed to add KCM tickets to table.\n");
652         talloc_zfree(renew_tgt_ctx);
653         return;
654     }
```


> It might be better to move this to a function on its own, something like:
> 
> ```c
> errno_t kcm_renewals_init(...)
> {
> #ifndef HAVE_KCM_RENEWAL
>     return EOK;
> #else
>    do stuff
> #endif
> }
> ```

Can you help me understand what is the benefit of this change?

pbrezina commented:
> Thanks a lot Pavel for the further review, it is better to get everything resolved now than having to fix issues later.
> 
> > If the cache contains uid that is not resolvable then kcm fails to start:
> > ```
> > [root /var/log/sssd]# /usr/libexec/sssd/sssd_kcm --uid 0 --gid 0 --debug-level 0xfff0
> > (2021-03-25 12:12:49:260824): [sssd] [become_user] (0x0200): Trying to become user [0][0].
> > (2021-03-25 12:12:49:260883): [sssd] [become_user] (0x0200): Already user [0].
> > (2021-03-25 12:12:49:263412): [kcm] [ldb] (0x0400): server_sort:Unable to register control with rootdse!
> > (2021-03-25 12:12:49): [kcm] [server_setup] (0x0040): Starting with debug level = 0xfff0
> > (2021-03-25 12:12:49): [kcm] [server_setup] (0x0400): CONFDB: /var/lib/sss/db/config.ldb
> > (2021-03-25 12:12:49): [kcm] [kcm_get_ccdb_be] (0x0100): KCM database type: secdb
> > (2021-03-25 12:12:49): [kcm] [kcm_ccdb_init] (0x0200): KCM back end: libsss_secrets
> > (2021-03-25 12:12:49): [kcm] [ccdb_secdb_init] (0x2000): secdb initialized
> > (2021-03-25 12:12:49): [kcm] [sss_sec_list_cc_uids] (0x2000): uid: [91600000]
> > (2021-03-25 12:12:49): [kcm] [sss_sec_list_cc_uids] (0x2000): uid: [1000]
> > (2021-03-25 12:12:49): [kcm] [ccdb_secdb_renew_init] (0x2000): Found [2] ccache uids
> > (2021-03-25 12:12:49): [kcm] [renew_check_ccaches] (0x0040): Failed to get pwd entry for [91600000]
> > (2021-03-25 12:12:49): [kcm] [ccdb_secdb_renew_init] (0x0040): Error checking ccaches in secdb
> > (2021-03-25 12:12:49): [kcm] [kcm_ccdb_renew_init] (0x0020): Failure to execute ccdb renewal init
> > (2021-03-25 12:12:49): [kcm] [kcm_process_init] (0x0010): fatal error initializing KCM ccdb renewals
> > (2021-03-25 12:12:49): [kcm] [kcm_responder_ctx_destructor] (0x0400): Responder is being shut down
> > ```
> 
> Is it valid to treat renewal failures as not fatal, and return EOK from `kcm_ccdb_renew_init` instead of current behavior shown below? If not what is the preferred way to handle this?

Fail to setup renewals should be fatal. But invalid uid inside the cache is not fatal, it may happen quite easily -- user was removed from LDAP and then from SSSD cache but kdestroy was not called before the user was removed. In the future we should probably purge such ccaches.

> > It might be better to move this to a function on its own, something like:
> > ```c
> > errno_t kcm_renewals_init(...)
> > {
> > #ifndef HAVE_KCM_RENEWAL
> >     return EOK;
> > #else
> >    do stuff
> > #endif
> > }
> > ```
> 
> Can you help me understand what is the benefit of this change?

The benefit is that the logic is confined to the function and not to its caller. The current code requires #ifdefs on two places. If you move it to a separate function, you only need one #ifdef and the caller does not have to care about the support. It is not that important at this place, but it is generally a good practice. The caller should know how to call the function not when to call it.

justin-stephenson commented on src/responder/kcm/kcm.c:
```diff
@@ -257,6 +260,7 @@ static int kcm_process_init(TALLOC_CTX *mem_ctx,
         goto fail;
     }
 
+#ifdef HAVE_KCM_RENEWAL
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcmsrv_ccache_secdb.c:
```diff
@@ -577,6 +580,211 @@ static errno_t ccdb_secdb_init(struct kcm_ccdb *db,
+    struct passwd *pwd;
+    errno_t ret;
+
+	cli_cred = talloc_zero(ev, struct cli_creds);
```
Fixed.

justin-stephenson commented:
> Hi Justin,
> I'm sorry it takes so long, but there are few more thinks. They are mostly minor, but one makes kcm fail to start.
> 
These issues are addressed in the latest version.

pbrezina commented:
Sorry, it still does not work correctly. Right now you don't continue with next user if you did not find the current. You want:

```diff
diff --git a/src/responder/kcm/kcmsrv_ccache_secdb.c b/src/responder/kcm/kcmsrv_ccache_secdb.c
index 7591c5687..14315e01b 100644
--- a/src/responder/kcm/kcmsrv_ccache_secdb.c
+++ b/src/responder/kcm/kcmsrv_ccache_secdb.c
@@ -707,7 +707,7 @@ static errno_t renew_check_ccaches(struct krb5_ctx *krb5_ctx,
     struct passwd *pwd;
     errno_t ret;
 
-    cli_cred = talloc_zero(ev, struct cli_creds);
+    cli_cred = talloc_zero(NULL, struct cli_creds);
     if (cli_cred == NULL) {
         return ENOMEM;
     }
@@ -716,11 +716,10 @@ static errno_t renew_check_ccaches(struct krb5_ctx *krb5_ctx,
     for (int i = 0; i < uid_count; i++) {
         pwd = getpwuid(uids[i]);
         if (pwd == NULL) {
-            talloc_zfree(cli_cred);
             DEBUG(SSSDBG_OP_FAILURE, "Failed to get pwd entry for [%d]\n",
                                      uids[i]);
             /* Not fatal */
-            return EOK;
+            continue;
         }
 
         cli_cred->ucred.uid = pwd->pw_uid;
@@ -728,16 +727,14 @@ static errno_t renew_check_ccaches(struct krb5_ctx *krb5_ctx,
 
         ret = secdb_container_url_req(state, sctx, cli_cred, &sreq);
         if (ret != EOK) {
-            talloc_zfree(cli_cred);
-            return EINVAL;
+            goto done;
         }
 
         ret = sss_sec_list(state, sreq, &keys, &nkeys);
         if (ret == ENOENT) {
             nkeys = 0;
         } else if (ret != EOK) {
-            talloc_zfree(cli_cred);
-            return EINVAL;
+            goto done;
         }
         DEBUG(SSSDBG_TRACE_INTERNAL, "Found [%zu] ccaches under uid [%u]\n",
                                      nkeys, uids[i]);
@@ -746,14 +743,16 @@ static errno_t renew_check_ccaches(struct krb5_ctx *krb5_ctx,
         for (size_t j = 0; j < nkeys; j++) {
             ret = renew_check_creds(krb5_ctx, state, sctx, cli_cred, keys[j]);
             if (ret != EOK && ret != ENOENT) {
-                talloc_zfree(cli_cred);
                 DEBUG(SSSDBG_OP_FAILURE, "Failed to check credentials\n");
-                return EINVAL;
+                goto done;
             }
         }
     }
 
-    return EOK;
+done:
+    talloc_free(cli_cred);
+
+    return ret;
 }
 
 static errno_t ccdb_secdb_renew_init(struct krb5_ctx *krb5_ctx,
```

Also what do we need the renew hash table for? I don't see where you take advantage of it? You always call `kcm_ccdb_renew_init` from `kcm_renew_tgt_timer_handler` so you always iterate over all cached credentials before you attempt to renew it. What was the intention here?

pbrezina commented:
> > Also what do we need the renew hash table for? I don't see where you take advantage of it? You always call `kcm_ccdb_renew_init` from `kcm_renew_tgt_timer_handler` so you always iterate over all cached credentials before you attempt to renew it. What was the intention here?
> 
> i was using krb5 renewals as a starting point for my code, `add_tgt_to_renew_table` uses the hash table. Is it better to just use an array of `renew_data` structures? `kcm_renew_all_tgts` is where we retrieve the hash table entries to check for renewals.

In this case the hash table is completely redundant. I think we have two options:

1. Keep the hash table. But in this case it would be better to maintain it from store/init/destroy/remove operations instead of populating it just before the refresh.
2. Drop the hash table and make the code much more straightforward. You don't need to renew_init on process start but only from the timer. You just want to iterate over all ccaches and find renewable tgts in the timer handler. (I favor this as long as there won't be performance issues since it is much easier)

In both cases, you'll probably end up with renew_init unneeded and you will gain support for all backends not just ccdb.

justin-stephenson commented:
> In this case the hash table is completely redundant. I think we have two options:
> 
>     1. Keep the hash table. But in this case it would be better to maintain it from store/init/destroy/remove operations instead of populating it just before the refresh.
> 
>     2. Drop the hash table and make the code much more straightforward. You don't need to renew_init on process start but only from the timer. You just want to iterate over all ccaches and find renewable tgts in the timer handler. (I favor this as long as there won't be performance issues since it is much easier)
> 
> 
> In both cases, you'll probably end up with renew_init unneeded and you will gain support for all backends not just ccdb.

Thank you, I will work on revising the PR for option 2. Just for information I am out on PTO starting April 2nd and returning April 12th.

pbrezina reviewed (changes requested):
Hi Justin. Thank you for the changes, the design looks good now. Its nice to see how the amount of code got reduced. There are lots of talloc issues that needs to be fixed, otherwise just few minor comments.

I think most of the talloc issues were in older code, the newer one was better in this area. You can check https://sssd.io/contrib/talloc.html again to get more information about it.

Please, also revisit the debug messages, make them more describing and if ret is available, print its value like this: `[%d]: %s\n, ret, sss_strerror(ret)`

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -313,4 +346,399 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+                goto done;
+            }
+
+            tevent_schedule_immediate(imm, renew_tgt_ctx->ev, kcm_renew_tgt,
```
This was just how it was done in `krb5_renew_tgt.c`

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -313,4 +346,399 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+    if (tgtt.renew_till >= tgtt.endtime && tgtt.renew_till >= now
+        && tgtt.endtime >= now && start_renew <= now) {
+            DEBUG(SSSDBG_TRACE_INTERNAL, "Renewal cred ready!\n");
+            auth_data = talloc_zero(renew_tgt_ctx, struct auth_data);
```
When I make this change the renewal does not occur, in the logs

[kcm_creds_check_times] (0x2000): Renewal cred ready!

justin-stephenson replied:
See my reply to your other memory leak auth_data comment, I left this as-is with the approach taken.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -313,4 +346,399 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+    struct kcm_renew_tgt_ctx *ctx = NULL;
+    errno_t ret;
+
+    ctx = talloc_zero(auth_data, struct kcm_renew_tgt_ctx);
```
I tried the first option but no luck...

~~~
diff --git a/src/responder/kcm/kcm_renew.c b/src/responder/kcm/kcm_renew.c
index 71a0bfbdf..a5f23e1ba 100644
--- a/src/responder/kcm/kcm_renew.c
+++ b/src/responder/kcm/kcm_renew.c
@@ -448,7 +448,7 @@ static void kcm_renew_tgt(struct tevent_context *ev,
 
     auth_data = talloc_get_type(private_data, struct auth_data);
 
-    ctx = talloc_zero(auth_data, struct kcm_renew_tgt_ctx);
+    ctx = talloc_zero(auth_data->krb5_ctx, struct kcm_renew_tgt_ctx);
     if (ctx == NULL) {
         DEBUG(SSSDBG_FATAL_FAILURE, "ctx talloc_zero failed\n");
         return;
@@ -531,7 +531,7 @@ static errno_t kcm_creds_check_times(TALLOC_CTX *mem_ctx,
     if (tgtt.renew_till >= tgtt.endtime && tgtt.renew_till >= now
         && tgtt.endtime >= now && start_renew <= now) {
             DEBUG(SSSDBG_TRACE_INTERNAL, "Renewal cred ready!\n");
-            auth_data = talloc_zero(renew_tgt_ctx, struct auth_data);
+            auth_data = talloc_zero(mem_ctx, struct auth_data);
             if (auth_data == NULL) {
                 DEBUG(SSSDBG_CRIT_FAILURE, "talloc_zero failed.\n");
                 ret = ENOMEM;
~~~

justin-stephenson replied:
Okay thanks, I went with the second approach and I am now also freeing auth_data in kcm_renew_tgt_done() - please check it.

justin-stephenson commented on src/responder/kcm/kcmsrv_ccache_secdb.c:
```diff
@@ -818,10 +822,176 @@ static errno_t ccdb_secdb_get_default_recv(struct tevent_req *req,
+
+    DEBUG(SSSDBG_TRACE_INTERNAL, "Found [%lu] ccache uuids\n", uuid_list_count);
+
+    cc_list = talloc_zero_array(tmp_ctx, struct kcm_ccache *, uuid_list_count + 1);
```
When I move this line inside ccdb_secdb_get_cc_for_uuid I get a crash but I can't figure out what i'm doing wrong

http://pastebin.test.redhat.com/961890

pbrezina replied:
Well, you didn't make it an output variable :-) Here you go:

```diff
diff --git a/src/responder/kcm/kcmsrv_ccache_secdb.c b/src/responder/kcm/kcmsrv_ccache_secdb.c
index ee3601458..9c844beca 100644
--- a/src/responder/kcm/kcmsrv_ccache_secdb.c
+++ b/src/responder/kcm/kcmsrv_ccache_secdb.c
@@ -822,12 +822,12 @@ static errno_t ccdb_secdb_get_default_recv(struct tevent_req *req,
     return EOK;
 }
 
-static errno_t ccdb_secdb_get_cc_for_uuid(size_t uuid_list_count,
+static errno_t ccdb_secdb_get_cc_for_uuid(TALLOC_CTX *mem_ctx,
+                                          size_t uuid_list_count,
                                           const char **uuid_list,
                                           const char **uid_list,
                                           struct ccdb_secdb *secdb,
-                                          struct kcm_ccache **cc_list,
-                                          int *_new_count)
+                                          struct kcm_ccache ***_cc_list)
 {
     TALLOC_CTX *tmp_ctx;
     errno_t ret;
@@ -837,6 +837,7 @@ static errno_t ccdb_secdb_get_cc_for_uuid(size_t uuid_list_count,
     char *uuid_str;
     char *secdb_key;
     struct cli_creds cli_cred;
+    struct kcm_ccache **cc_list;
     int real_count = 0;
 
     tmp_ctx = talloc_new(NULL);
@@ -845,6 +846,13 @@ static errno_t ccdb_secdb_get_cc_for_uuid(size_t uuid_list_count,
         return ret;
     }
 
+    cc_list = talloc_zero_array(tmp_ctx, struct kcm_ccache *,
+                                uuid_list_count + 1);
+    if (cc_list == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
     for (size_t i = 0; i < uuid_list_count; i++) {
         struct passwd *pwd;
 
@@ -900,7 +908,8 @@ static errno_t ccdb_secdb_get_cc_for_uuid(size_t uuid_list_count,
             goto done;
         }
 
-        ret = secdb_get_cc(cc_list, secdb->sctx, secdb_key, &cli_cred, &cc_list[real_count]);
+        ret = secdb_get_cc(cc_list, secdb->sctx, secdb_key, &cli_cred,
+                           &cc_list[real_count]);
         if (ret != EOK) {
             DEBUG(SSSDBG_CRIT_FAILURE, "Failed to get ccache [%d]: %s\n",
                                        ret, sss_strerror(ret));
@@ -920,7 +929,13 @@ static errno_t ccdb_secdb_get_cc_for_uuid(size_t uuid_list_count,
 
     cc_list = talloc_realloc(tmp_ctx, cc_list, struct kcm_ccache *,
                              real_count + 1);
-    *_new_count = real_count;
+    if (cc_list == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    cc_list[real_count] = NULL;
+    *_cc_list = talloc_steal(mem_ctx, cc_list);
 
     return EOK;
 done:
@@ -944,8 +959,6 @@ static errno_t ccdb_secdb_list_all_cc(TALLOC_CTX *mem_ctx,
     const char **uid_list;
     const char **uuid_list;
     size_t uuid_list_count;
-    struct kcm_ccache **cc_list;
-    int new_count = 0;
 
     DEBUG(SSSDBG_TRACE_INTERNAL, "Retrieving all ccaches\n");
 
@@ -966,24 +979,15 @@ static errno_t ccdb_secdb_list_all_cc(TALLOC_CTX *mem_ctx,
 
     DEBUG(SSSDBG_TRACE_INTERNAL, "Found [%lu] ccache uuids\n", uuid_list_count);
 
-    cc_list = talloc_zero_array(tmp_ctx, struct kcm_ccache *, uuid_list_count + 1);
-    if (cc_list == NULL) {
-        ret = ENOMEM;
-        goto done;
-    }
-
     /* New count is full cc list size minus getpwuid() failures */
-    ret = ccdb_secdb_get_cc_for_uuid(uuid_list_count, uuid_list, uid_list, secdb,
-                                     cc_list, &new_count);
+    ret = ccdb_secdb_get_cc_for_uuid(mem_ctx, uuid_list_count, uuid_list,
+                                     uid_list, secdb, _cc_list);
     if (ret != EOK) {
         DEBUG(SSSDBG_TRACE_INTERNAL, "Error getting cc list from uuid list "
                                      "[%d]: %s\n", ret, sss_strerror(ret));
         goto done;
     }
 
-    cc_list[new_count] = NULL;
-    *_cc_list = talloc_steal(mem_ctx, cc_list);
-
     DEBUG(SSSDBG_TRACE_FUNC, "Retrieving all caches done\n");
     ret = EOK;
```

justin-stephenson commented on src/responder/kcm/kcm.c:
```diff
@@ -48,6 +50,39 @@ static int kcm_responder_ctx_destructor(void *ptr)
+
+    ret = kcm_get_renewal_config(kctx, &krb5_ctx, &renew_intv);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "fatal error getting KCM renewal config\n");
```
Is printing the ret and error string necessary for simple malloc functions like...

~~~
373 
374     krreq = talloc_zero(mem_ctx, struct krb5child_req);
375     if (krreq == NULL) {
376         ret = ENOMEM;
~~~

And also what about `kerr` error codes, is there a utility function to print the kerr string similar to `ldb_strerror` ?
~~~
611     kerr = krb5_init_context(&krb_context);
612     if (kerr != 0) {
613         ret = ENOMEM;
614         DEBUG(SSSDBG_CRIT_FAILURE, "Failed to init krb5 context\n");
615         goto done;
616     }
~~~

pbrezina replied:
> Is printing the ret and error string necessary for simple malloc functions like...

No, in this case ret has no meaning (you set it yourself). It is enough to print it only when you consume its value (ie when it is returned from called function).

> And also what about kerr error codes, is there a utility function to print the kerr string similar to ldb_strerror ?

`sss_krb5_get_error_message`, it requires also `sss_krb5_free_error_message`

justin-stephenson commented on src/util/secrets/secrets.c:
```diff
@@ -937,6 +939,93 @@ static char *local_dn_to_path(TALLOC_CTX *mem_ctx,
+                              struct sss_sec_ctx *sec,
+                              const char ***_uuid_list,
+                              const char ***_uid_list,
+                              size_t *_uuid_list_count)
```
> Hi Justin. Thank you for the changes, the design looks good now. Its nice to see how the amount of code got reduced. There are lots of talloc issues that needs to be fixed, otherwise just few minor comments.

Thanks very much for the detailed review. I just pushed a new version which should address each review item aside from 1 or 2 items - for those I replied in the conversation for further clarification.

pbrezina commented on src/responder/kcm/kcmsrv_ccache_secdb.c:
```diff
@@ -818,10 +822,176 @@ static errno_t ccdb_secdb_get_default_recv(struct tevent_req *req,
+        pwd = getpwuid(uid);
+        if (pwd == NULL) {
+            ret = errno;
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to get pwd entry for "
```
SSSDBG_MINOR_FAILURE since it is not impotant. It will not also invoke the new backtrace functionality.

pbrezina replied:
Perhaps also change the message to something more useful for users like "Unable to resolve user [uid] who is the owner of an existing ccache [%d]: %s\n"

justin-stephenson replied:
Fixed.

pbrezina commented on src/man/sssd-kcm.8.xml:
```diff
@@ -162,6 +162,42 @@ systemctl restart sssd-kcm.service
+        </para>
+            <programlisting>
+tgt_renewal = true
+tgt_renewal_inherit = testad.domain
```
use generic "domain-name" instead of testad.domain

justin-stephenson replied:
Fixed.

pbrezina commented on src/responder/kcm/kcm.c:
```diff
@@ -48,6 +50,43 @@ static int kcm_responder_ctx_destructor(void *ptr)
+                                    ret, sss_strerror(ret));
+        return ret;
+    }
+
```
It might be good to print the config here so we can see it in the logs.

justin-stephenson replied:
print the config? Sorry can you elaborate

pbrezina replied:
Print the option values that you read.

justin-stephenson replied:
I printed the values being set in `kcm_set_options`

justin-stephenson replied:
Fixed.

pbrezina reviewed (changes requested):
Thank you. We're almost there, see minor comments.

pbrezina commented:
And one more thing - add [release notes](https://github.com/SSSD/sssd/blob/master/.git-commit-template#L7) and upstream ticket to the commit messages, e.g.
* https://github.com/SSSD/sssd/pull/5610/commits/c556e2543148792e05f435d25458f5309be74838
* https://github.com/SSSD/sssd/pull/5610/commits/3973e3f231dbe4c344ded9d4fdbf3effd3ae32be

justin-stephenson commented:
> And one more thing - add [release notes](https://github.com/SSSD/sssd/blob/master/.git-commit-template#L7) and upstream ticket to the commit messages, e.g.
> 
>     * [c556e25](https://github.com/SSSD/sssd/commit/c556e2543148792e05f435d25458f5309be74838)
> 
>     * [3973e3f](https://github.com/SSSD/sssd/commit/3973e3f231dbe4c344ded9d4fdbf3effd3ae32be)

Should `:feature:` be added to all commits? Or just the main one implementing the feature

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -324,4 +358,398 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        ret = ENOMEM;
```
Fixed.

justin-stephenson commented on src/responder/kcm/kcm_renew.c:
```diff
@@ -324,4 +358,398 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
+        goto done;
+    }
+
+    if (cc_list == NULL) {
```
Fixed.

pbrezina commented:
> > And one more thing - add [release notes](https://github.com/SSSD/sssd/blob/master/.git-commit-template#L7) and upstream ticket to the commit messages, e.g.
> > ```
> > * [c556e25](https://github.com/SSSD/sssd/commit/c556e2543148792e05f435d25458f5309be74838)
> > 
> > * [3973e3f](https://github.com/SSSD/sssd/commit/3973e3f231dbe4c344ded9d4fdbf3effd3ae32be)
> > ```
> 
> Should `:feature:` be added to all commits? Or just the main one implementing the feature

Just once. Each release note should be there only once.

justin-stephenson commented:
> And one more thing - add [release notes](https://github.com/SSSD/sssd/blob/master/.git-commit-template#L7) and upstream ticket to the commit messages, e.g.
> 
>     * [c556e25](https://github.com/SSSD/sssd/commit/c556e2543148792e05f435d25458f5309be74838)
> 
>     * [3973e3f](https://github.com/SSSD/sssd/commit/3973e3f231dbe4c344ded9d4fdbf3effd3ae32be)

Done, please check it.

pbrezina commented on src/responder/kcm/kcmsrv_ccache_secdb.c:
```diff
@@ -818,10 +822,181 @@ static errno_t ccdb_secdb_get_default_recv(struct tevent_req *req,
+        goto done;
+    }
+
+
```
two empty lines instead of one

justin-stephenson replied:
Fixed.

alexey-tikhonov replied:
Do we actually have anything in the coding style on this matter?
I'm asking because I also like newlines :)

pbrezina replied:
I don't think there is specific statement that we can't use more then two lines. Why would you like to use more than one?

pbrezina commented:
Thank you for your patience, last two nitpick in code and we can push.

And perhaps different wording for the release notes (fill in the version):
```
    :feature: Added support for automatic renewal of renewable TGTs that are stored in
    KCM ccache. This can be enabled by setting `tgt_renewal = true`. See the sssd-kcm man page for
    more details. This feature requires MIT Kerberos X.Y.Z
```

justin-stephenson commented:
> Thank you for your patience, last two nitpick in code and we can push.
> 
> And perhaps different wording for the release notes (fill in the version):
> 
> ```
>     :feature: Added support for automatic renewal of renewable TGTs that are stored in
>     KCM ccache. This can be enabled by setting `tgt_renewal = true`. See the sssd-kcm man page for
>     more details. This feature requires MIT Kerberos X.Y.Z
> ```

Fixed and pushed, not sure if ``MIT Kerberos krb5-1.19-0.beta2.3 or higher.`` is the correct wording however.

pbrezina commented:
Thank you. Ack.

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/5450

* `master`
    * ec932d35172819ac68343355faaad4dc6ffae688 - KCM: Disable responder idle timeout with renewals
    * ddcedbf3bc6b267d40d8a7edcb65f8d61ec13dd1 - KCM: Conditionally build KCM renewals support
    * 0202eb53ab18b5eeac53fc96bf5e0569276e3767 - INTG: Add KCM Renewal integration test
    * a55405b3edd6312a5e39567e4bdde5522ffc6a0a - TESTS: Add kcm_renewals unit test
    * 1dc3c33c8d2f4ca4a41b186746c44f74510c2f38 - SECRETS: Don't hardcode SECRETS_DB_PATH
    * 599f0ad056dc8fc052395d5abe0e110e4e68a886 - KCM: Prepare and execute renewals
    * 993b66d48d555c59e619d7ef3b494248a82587ac - KCM: Read and set KCM renewal and krb5 options
